pkg/tfbridge/detailed_diff_test.go

```diff
@@ -0,0 +1,1631 @@
+
+	require.Equal(t, makeBaseDiff(nilVal, nilVal), NoDiff)
+	require.Equal(t, makeBaseDiff(nilVal, nilVal), NoDiff)
+	require.Equal(t, makeBaseDiff(nilVal, nonNilVal), Add)
+	require.Equal(t, makeBaseDiff(nilVal, nonNilVal), Add)
+	require.Equal(t, makeBaseDiff(nonNilVal, nilVal), Delete)
+	require.Equal(t, makeBaseDiff(nonNilVal, nilVal), Delete)
+	require.Equal(t, makeBaseDiff(nonNilVal, nilArr), Delete)
+	require.Equal(t, makeBaseDiff(nonNilVal, nilMap), Delete)
+	require.Equal(t, makeBaseDiff(nonNilVal, nonNilVal2), Undecided)
+}
+
+func TestMakePropDiff(t *testing.T) {
+	tests := []struct {
+		name string
+		old  resource.PropertyValue
+		new  resource.PropertyValue
+		etf  shimschema.Schema
+		eps  *SchemaInfo
+		want *pulumirpc.PropertyDiff
+	}{
+		{
+			name: "unchanged non-nil",
+			old:  resource.NewStringProperty("same"),
+			new:  resource.NewStringProperty("same"),
+			want: nil,
+		},
+		{
+			name: "unchanged nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewNullProperty(),
+			want: nil,
+		},
+		{
+			name: "unchanged not present",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewNullProperty(),
+			want: nil,
+		},
+		{
+			name: "added",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD},
+		},
+		{
+			name: "deleted",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE},
+		},
+		{
+			name: "changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE},
+		},
+		{
+			name: "changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD},
+		},
+		{
+			name: "changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE},
+		},
+		{
+			name: "tf force new unchanged",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("old"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: nil,
+		},
+		{
+			name: "tf force new changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE},
+		},
+		{
+			name: "tf force new changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE},
+		},
+		{
+			name: "tf force new changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
```

Comment:
This implies to me that we don't distinguish between a null property value and no property value. Do we consider these the same?

```go
resource.PropertyMap{"null": resource.NewNullProperty()}
```

```go
resource.PropertyMap{}
```

If the above maps are not the same, then this should be an update ("old" -> null). If the above maps are the same, then I have more questions:

Are these maps the same?
```go
resource.PropertyMap{"empty": resource.PropertyMap{"null": resource.NewNullProperty()}}
```
```go
resource.PropertyMap{}
```

This gets even more confusing when we introduce arrays: `[null, "non-null"]` is representable in TF.

Reply:
TFPF definitely distinguishes between `null` and not present. I believe that SDKv2 does as well in some cases, which is why I'm concerned about treating null and missing as identical. This becomes especially confusing when we see collections with `null`.

Reply:
Can you give an example of the SDKv2 treating null and undefined differently? I have not been able to find any so far

Reply:
In HCL, null and undefined are the same: https://developer.hashicorp.com/terraform/language/expressions/types#null

Reply:
I originally started this PR with handling of null vs undefined but found no uses for that.

Reply:
I've covered these cases and more with cross-tests in https://github.com/pulumi/pulumi-terraform-bridge/pull/2405/commits/c55a0921406e71f106091790fd6246719519305f

LMK if there are any more concerns here.

Reply:
This code needs to work for PF ...

Reply:
Yeah, I am aware - let's solve for PF when we have tests for it, etc.

The current PR is for SDKv2

Reply:
Edit: not 100% true - there might be cases like Ian mentioned where it does distinguish between some of that. Found an issue with the tests, will investigate more: https://github.com/pulumi/pulumi-terraform-bridge/pull/2405/commits/051d1a73b258d777086ca19d470da75d34dccf0e

Reply:
Okay, so I believe null vs unspecified is not a diff in all cases for the SDKv2.

There are cases where empty collections are treated differently than null ones but I think that was not part of the discussion here. For reference the empty collection behaviour is covered in cross-tests in https://github.com/pulumi/pulumi-terraform-bridge/pull/2405/files#diff-5ecbec35cb8a03b89c5be410138532ea2da5904bfd52fdcb604d3b75414a6044R1044